File: src/main/java/com/lcaohoanq/nocket/domain/auth/ForgotPasswordService.java
package com.lcaohoanq.nocket.domain.auth;

import com.lcaohoanq.nocket.constant.EmailSubject;
import com.lcaohoanq.nocket.enums.EmailCategoriesEnum;
import com.lcaohoanq.nocket.domain.otp.Otp;
import com.lcaohoanq.nocket.domain.user.User;
import com.lcaohoanq.nocket.domain.mail.IMailService;
import com.lcaohoanq.nocket.domain.otp.IOtpService;
import com.lcaohoanq.nocket.util.OtpUtil;
import jakarta.mail.MessagingException;
import java.time.LocalDateTime;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.thymeleaf.context.Context;

@Slf4j
@Service
@RequiredArgsConstructor
public class ForgotPasswordService implements IForgotPasswordService {

    private final IMailService mailService;
    private final IOtpService otpService;

    @Override
    @Transactional
    public void sendEmailOtp(User existingUser) throws MessagingException {
        Context context = new Context();
        String otp = OtpUtil.generateOtp();
        context.setVariable("name", existingUser.getName());
        context.setVariable("otp", otp);

        mailService.sendMail(existingUser.getEmail(),
                             EmailSubject.subjectForgotPassword(existingUser.getName()),
                             EmailCategoriesEnum.FORGOT_PASSWORD.getType(),
                             context);

        Otp otpEntity = new Otp();
        otpEntity.setOtp(otp);
        otpEntity.setEmail(existingUser.getEmail());
        otpEntity.setExpiredAt(LocalDateTime.now().plusMinutes(5));
        otpEntity.setUsed(false);
        otpEntity.setExpired(false);
        
        otpService.createOtp(otpEntity);
    }
}

File: src/main/kotlin/com/lcaohoanq/nocket/domain/auth/ForgotPasswordService.kt
package com.lcaohoanq.nocket.domain.auth

import com.lcaohoanq.nocket.constant.EmailSubject
import com.lcaohoanq.nocket.domain.mail.IMailService
import com.lcaohoanq.nocket.domain.otp.IOtpService
import com.lcaohoanq.nocket.domain.otp.Otp
import com.lcaohoanq.nocket.domain.user.User
import com.lcaohoanq.nocket.enums.EmailCategoriesEnum
import com.lcaohoanq.nocket.util.OtpUtil
import jakarta.mail.MessagingException
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.thymeleaf.context.Context
import java.time.LocalDateTime

@Slf4j
@Service
open class ForgotPasswordService(
    private val mailService: IMailService,
    private val otpService: IOtpService
) : IForgotPasswordService {
    
    @Transactional
    @Throws(MessagingException::class)
    override fun sendEmailOtp(existingUser: User) {
        val context = Context()
        val otp = otpService.generateOtp()
        context.setVariable("name", existingUser.name)
        context.setVariable("otp", otp)

        mailService.sendMail(
            existingUser.email,
            EmailSubject.subjectForgotPassword(existingUser.name),
            EmailCategoriesEnum.FORGOT_PASSWORD.type,
            context
        )

        val otpEntity = Otp()
        otpEntity.otp = otp
        otpEntity.email = existingUser.email
        otpEntity.expiredAt = LocalDateTime.now().plusMinutes(5)
        otpEntity.isUsed = false
        otpEntity.isExpired = false

        otpService.createOtp(otpEntity)
    }
}
